keps/sig-node/20191025-image-pull-progress.md

```diff
@@ -0,0 +1,402 @@
+overall usability is not satisfying at all.
+
+To improve the amount of information when calling `kubectl get pods`, we will
+now take the thrown events into account as well. To achieve that and stay
+backwards compatible, a new output variant `events` will be added to the
+`kubectl get` flag `-o/--output`. If the flag is set, `kubectl` will map
+(manually implemented) relevant information into the default (table) output in a
+similar fashion to the `-o wide`. For example, the output for pods pulling a
+container image will now look like:
+
+```
+> kubectl get pods -o events
+NAMESPACE     NAME                       READY   STATUS              RESTARTS   AGE   EVENT
+kube-system   coredns-65c865486d-s2fs5   0/1     ContainerCreating   0          3s    Pulling Image (42%)
+```
+
+### Implementation Details/Notes/Constraints
+
+The major caveat for this implementation approach is the increased amount of
+thrown events from the kubelet during an image pull. This should be limited in a
+fashion that the API surface is not strongly impacted.
+
+### Risks and Mitigations
+
+The interface changes to `kubectl` should increase the usability to the users,
+whereas it opens the doors for future correlations between events and
+`kubectl get` requests. There is a risk that the current design proposal does
+not fulfill that need.
+
+## Production Readiness Review Questionnaire
+
+### Feature enablement and rollback
+
+- **How can this feature be enabled / disabled in a live cluster?**
+
+  Via the kubelet dynamic configuration feature, where we will add a dedicated
+  option for it.
+
+- **Can the feature be disabled once it has been enabled (i.e., can we roll
+  back the enablement)?**
+
+  Yes, but not during a image pull (already established gRPC stream).
+
+- **Will enabling / disabling the feature require downtime for the control
+  plane?**
+
+  No
+
+- **Will enabling / disabling the feature require downtime or reprovisioning
+  of a node?**
+
+  No
+
+- **What happens if a cluster with this feature enabled is rolled back? What
+  happens if it is subsequently upgraded again?**
+
+  Nothing, since the configuration option (enabled or disabled) is only a
+  toggle and relies on no other state.
+
+- **Are there tests for this?**
+
+  Yes, end to end tests for the kubelet.
+
+### Scalability
+
+- **Will enabling / using the feature result in any new API calls?
+  Describe them with their impact keeping in mind the supported limits
+  (e.g. 5000 nodes per cluster, 100 pods/s churn) focusing mostly on:**
+
+  - **components listing and/or watching resources they didn't before**
+  - **API calls that may be triggered by changes of some Kubernetes
+    resources (e.g. update object X based on changes of object Y)**
+  - **periodic API calls to reconcile state (e.g. periodic fetching state,
+    heartbeats, leader election, etc.)**
+
+  Yes, we throw new events during the image pull process. The amount can be
```

Comment:
s/amount/frequency/ would probably be better word

Reply:
Fixed :+1: